pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	return (secret != nil), nil
+}
+
+// DeleteSecret deletes a secret from ngrok by its reference.
+func (c *client) DeleteSecret(ctx context.Context, ref esv1.PushSecretRemoteRef) error {
+	secret, err := c.getSecretByVaultNameAndSecretName(ctx, c.vaultName, ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		// If the secret or vault do not exist, we can consider it deleted.
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if secret == nil {
+		return nil
+	}
+
+	return c.secretsClient.Delete(ctx, secret.ID)
+}
+
+func (c *client) Validate() (esv1.ValidationResult, error) {
+	// Validate the client can list secrets with a timeout. If we
+	// can list secrets, we assume the client is valid(API keys, URL, etc.)
+	iter := c.secretsClient.List(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for iter.Next(ctx) {
+		return esv1.ValidationResultReady, nil
+	}
+
+	if iter.Err() != nil {
+		return esv1.ValidationResultError, fmt.Errorf("store is not allowed to list secrets: %w", iter.Err())
+	}
+
+	return esv1.ValidationResultReady, nil
+}
+
+func (c *client) GetSecret(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	// Implementation for getting a secret from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetSecretMap(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	// Implementation for getting a map of secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetAllSecrets(ctx context.Context, _ esv1.ExternalSecretFind) (map[string][]byte, error) {
+	// Implementation for getting all secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) Close(_ context.Context) error {
+	return nil
+}
+
+func (c *client) getVaultByName(ctx context.Context, name string) (*ngrok.Vault, error) {
+	iter := c.vaultClient.List(nil)
+	for iter.Next(ctx) {
+		vault := iter.Item()
+		if vault.Name == name {
+			return vault, nil
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+
+	return nil, errVaultDoesNotExist
+}
+
+func (c *client) getSecretByVaultAndName(ctx context.Context, vault ngrok.Vault, name string) (*ngrok.Secret, error) {
+	iter := c.secretsClient.List(nil)
+	for iter.Next(ctx) {
+		secret := iter.Item()
+		if secret.Vault.ID != vault.ID {
+			continue
+		}
+
+		if secret.Name == name {
+			return secret, nil
+		}
+	}
+
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+
+	return nil, fmt.Errorf("secret '%s' does not exist: %w", name, errVaultSecretDoesNotExist)
+}
+
+func (c *client) getSecretByVaultNameAndSecretName(ctx context.Context, vaultName, secretName string) (*ngrok.Secret, error) {
+	vault, err := c.getVaultByName(ctx, vaultName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get vault '%s' by name: %w", vaultName, err)
+	}
+
+	if vault == nil {
```

Comment:
There is no possibility for this to be nil looking at getVaultByName, which always returns an error.

Reply:
This should be fixed now